Support text 2.0

/cc @bodigrim

```
$ cabal build --constraint 'text >= 2.0'
src/Data/ByteString/Lazy/Encoding/Internal.hs:156:30: error:
    • Couldn't match type ‘Word16’ with ‘Word8’
      Expected type: Ptr Word8
        Actual type: Ptr Word16
    • In the first argument of ‘T.fromPtr’, namely ‘workspace’
      In a stmt of a 'do' block:
        t <- T.fromPtr workspace (fromIntegral n)
      In the expression:
        do let p' :: Ptr Char
               p' = castPtr p
               ....
           n <- f (bufL buf) 0
           t <- T.fromPtr workspace (fromIntegral n)
           return ([t], buf {bufL = 0, bufR = 0})
    |
156 |               t <- T.fromPtr workspace (fromIntegral n)
    |
```

As a Hackage trustee I made revisions to prohibit unbuildable configurations:
https://hackage.haskell.org/package/bytestring-encoding-0.1.0.0/revisions/
https://hackage.haskell.org/package/bytestring-encoding-0.1.1.0/revisions/

--- a/.github/workflows/build.yaml
+++ b/.github/workflows/build.yaml
@@ -49,6 +49,11 @@ jobs:
             resolver: 'lts-18.23'
             os: macos-latest
             flags: ''
+          - ghc: '8.10.7'
+            resolver: 'lts-18.23'
+            os: ubuntu-latest
+            flags: ''
+            text2: true
           - ghc: '9.0.2'
             resolver: 'nightly-2022-01-26'
             os: ubuntu-latest
@@ -78,9 +83,14 @@ jobs:
         run: rm -r ~/.stack/setup-exe-cache
 
       - name: Build
+        shell: bash
         run: |
           echo "resolver: ${{ matrix.resolver }}" > stack.yaml
+          if [[ "${{ matrix.text2 }}" = "true" ]]; then
+            echo "extra-deps: [text-2.0]" >> stack.yaml
+          fi
           stack build --test --no-run-tests --bench --no-run-benchmarks ${{ matrix.flags }}
+
       - name: Test
         run: stack test ${{ matrix.flags }}
 

--- a/package.yaml
+++ b/package.yaml
@@ -24,7 +24,7 @@ tested-with: 'GHC==7.10.3 GHC==8.0.2 GHC==8.2.2 GHC==8.4.4 GHC==8.6.5 GHC==8.8.4
 dependencies:
 - base >= 4.7 && < 5
 - bytestring
-- text
+- text <2.1
 
 library:
   source-dirs: src

--- a/src/Data/ByteString/Lazy/Encoding/Internal.hs
+++ b/src/Data/ByteString/Lazy/Encoding/Internal.hs
@@ -1,6 +1,7 @@
 {-# OPTIONS_GHC -Wall #-}
 {-# OPTIONS_HADDOCK hide #-}
 {-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE RecordWildCards #-}
 module Data.ByteString.Lazy.Encoding.Internal
@@ -15,6 +16,7 @@ import Control.Monad
 import qualified Data.ByteString as B
 import qualified Data.ByteString.Unsafe as B
 import qualified Data.ByteString.Lazy as BL
+import Data.Char (ord)
 import qualified Data.Text as T
 import qualified Data.Text.Foreign as T
 import qualified Data.Text.Lazy as TL
@@ -127,6 +129,23 @@ decodeWith Enc.TextEncoding{ .. } inBufSize outBufSize b = TL.fromChunks $ unsaf
                       moveBytes (q `plusPtr` bufR buf) p (bufferAvailable buf)
                       go (B.drop (bufferAvailable buf) b : bs) buf{ bufR = bufR buf + bufferAvailable buf }
 
+#if MIN_VERSION_text(2,0,0)
+      flushOutBuf :: CharBuffer -> ForeignPtr Word8 -> IO ([T.Text], CharBuffer)
+      flushOutBuf buf workspace
+        | isEmptyBuffer buf = return ([], buf{ bufL=0, bufR=0 })
+        | otherwise =
+            withForeignPtr workspace $ \workspace' ->
+            withBuffer buf $ \p -> do
+              let f !i !j
+                    | bufR buf <= i = return j
+                    | otherwise = do
+                        (c, i') <- readCharBufPtr (castPtr p) i
+                        j' <- writeUTF8 workspace' j c
+                        f i' j'
+              n <- f (bufL buf) 0
+              t <- T.fromPtr workspace' (fromIntegral n)
+              return ([t], buf{ bufL=0, bufR=0 })
+#else
       flushOutBuf :: CharBuffer -> ForeignPtr Word16 -> IO ([T.Text], CharBuffer)
       flushOutBuf buf workspace
         | isEmptyBuffer buf = return ([], buf{ bufL=0, bufR=0 })
@@ -156,8 +175,13 @@ decodeWith Enc.TextEncoding{ .. } inBufSize outBufSize b = TL.fromChunks $ unsaf
               n <- f (bufL buf) 0
               t <- T.fromPtr workspace' (fromIntegral n)
               return ([t], buf{ bufL=0, bufR=0 })
+#endif
 
+#if MIN_VERSION_text(2,0,0)
+      loop :: [B.ByteString] -> Buffer Word8 -> CharBuffer -> ForeignPtr Word8 -> IO [T.Text]
+#else
       loop :: [B.ByteString] -> Buffer Word8 -> CharBuffer -> ForeignPtr Word16 -> IO [T.Text]
+#endif
       loop bs inBuf outBuf workspace = do
         (bs', inBuf1) <- fillInBuf bs inBuf
         if isEmptyBuffer inBuf1 then do
@@ -191,5 +215,66 @@ decodeWith Enc.TextEncoding{ .. } inBufSize outBufSize b = TL.fromChunks $ unsaf
 
   inBuf <- newByteBuffer inBufSize ReadBuffer
   outBuf <- newCharBuffer outBufSize WriteBuffer
+#if MIN_VERSION_text(2,0,0)
+  workspace <- mallocForeignPtrArray (outBufSize * 4)
+#else
   workspace <- if charSize == 2 then newForeignPtr_ nullPtr else mallocForeignPtrArray (outBufSize * 2)
+#endif
   loop (BL.toChunks b) inBuf outBuf workspace
+
+
+#if MIN_VERSION_text(2,0,0)
+
+writeUTF8 :: Ptr Word8 -> Int -> Char -> IO Int
+writeUTF8 p i c = do
+  let x = ord c
+  if x <= 0x7F then do
+    pokeElemOff p i (fromIntegral x)
+    return $! i+1
+  else if x <= 0x07FF then do
+    let (c1,c2) = ord2 c
+    pokeElemOff p i c1
+    pokeElemOff p (i+1) c2
+    return $! i+2
+  else if x <= 0xFFFF then do
+    let (c1,c2,c3) = ord3 c
+    pokeElemOff p i c1
+    pokeElemOff p (i+1) c2
+    pokeElemOff p (i+2) c3
+    return $! i+3
+  else do
+    let (c1,c2,c3,c4) = ord4 c
+    pokeElemOff p i c1
+    pokeElemOff p (i+1) c2
+    pokeElemOff p (i+2) c3
+    pokeElemOff p (i+3) c4
+    return $! i+4
+
+-- -----------------------------------------------------------------------------
+-- UTF-8 primitives, lifted from Data.Text.Fusion.Utf8
+
+ord2   :: Char -> (Word8,Word8)
+ord2 c = assert (n >= 0x80 && n <= 0x07ff) (x1,x2)
+    where
+      n  = ord c
+      x1 = fromIntegral $ (n `shiftR` 6) + 0xC0
+      x2 = fromIntegral $ (n .&. 0x3F)   + 0x80
+
+ord3   :: Char -> (Word8,Word8,Word8)
+ord3 c = assert (n >= 0x0800 && n <= 0xffff) (x1,x2,x3)
+    where
+      n  = ord c
+      x1 = fromIntegral $ (n `shiftR` 12) + 0xE0
+      x2 = fromIntegral $ ((n `shiftR` 6) .&. 0x3F) + 0x80
+      x3 = fromIntegral $ (n .&. 0x3F) + 0x80
+
+ord4   :: Char -> (Word8,Word8,Word8,Word8)
+ord4 c = assert (n >= 0x10000) (x1,x2,x3,x4)
+    where
+      n  = ord c
+      x1 = fromIntegral $ (n `shiftR` 18) + 0xF0
+      x2 = fromIntegral $ ((n `shiftR` 12) .&. 0x3F) + 0x80
+      x3 = fromIntegral $ ((n `shiftR` 6) .&. 0x3F) + 0x80
+      x4 = fromIntegral $ (n .&. 0x3F) + 0x80
+
+#endif

